generators/browser-compatible-base/src/ast/AbstractWriter.ts

```diff
@@ -150,6 +150,24 @@ export class AbstractWriter {
         this.indentLevel--;
     }
 
+    public delimit(
+        nodes: AbstractAstNode[],
+        delimiter: string,
+        writeFunction: (node: AbstractAstNode | undefined) => void,
+        prefix: string = "",
+        suffix: string = ""
+    ): void {
+        this.write(prefix);
+        if (nodes.length > 0) {
+            writeFunction(nodes[0]);
+            for (let i = 1; i < nodes.length; i++) {
+                this.write(delimiter);
+                writeFunction(nodes[i]);
+            }
+        }
+        this.write(suffix);
+    }
```

Comment:
Nice, I feel like we've needed something like this for a while. Couple things we can do to improve the readability at the call-site though:

1. Let's use an in-lined request with named parameters.
2. I'd probably drop the `prefix` and `suffix` arguments to start. I feel like it's marginally more readable just to include those at the call-site so that users can clearly see where the beginning and end of a `( ... )` is.
3. Should we write the arguments across newlines and indent/dedent the inner values by default? We could also add a `newline` option to control that behavior if we prefer.

In combination, I imagine something like this:

```typescript
public delimit({
  nodes,
  writeFunc,
  delimiter,
  newline,
}: {
  nodes: AbstractAstNode[],
  writeFunc: (node: AbstractAstNode) => void,
  delimiter: string,
  newline?: boolean
}): void {
  if (nodes.length === 0) {
    return;
  }
  ...
}
```

With that said, I imagine we'll eventually need something like Go's writeArguments to handle no-op arguments well. It might be simpler to repurpose that method as a separate TypeScript utility instead (which uses ts.TypeLiteral instead of go.TypeInstantiation).

I'll leave it to you though - either way works!

Reply:
For the prefix suffix, I was mostly thinking about leaving those in for cases like a union type, where you could use this for `< Type1 | Type2 | Type3 >` with a delimiter of "|" and diamond brackets for the prefix/suffix. I can see it being more ergonomic to simplify it for the common case though